core/src/main/java/com/linecorp/armeria/internal/eventloop/EventLoopCheckingCompletableFuture.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.eventloop;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.Flags;
+
+/**
+ * A {@link CompletableFuture} that warns the user if they call a method that blocks the event loop.
+ */
+public final class EventLoopCheckingCompletableFuture<T> extends CompletableFuture<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(EventLoopCheckingCompletableFuture.class);
+
+    @Override
+    public T get() throws InterruptedException, ExecutionException {
+        maybeLogIfOnEventLoop();
+        return super.get();
+    }
+
+    @Override
+    public T get(long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        maybeLogIfOnEventLoop();
+        return super.get(timeout, unit);
+    }
+
+    @Override
+    public T join() {
+        maybeLogIfOnEventLoop();
+        return super.join();
+    }
+
+    private void maybeLogIfOnEventLoop() {
+        if (Flags.checkFutureJoinOnEventLoop() && Thread.currentThread() instanceof EventLoopThread) {
+            logger.warn("Calling a blocking method on CompletableFuture from an event loop thread. You " +
+                        "should never do this as this will usually result in significantly reduced " +
+                        "performance of the server, generally crippling its ability to handle high load, or " +
+                        "even result in deadlock which cannot be recovered from. Use " +
+                        "ServiceRequestContext.blockingExecutor to run this logic instead or switch to using " +
+                        "asynchronous methods like thenApply. If you really believe it is fine to block the " +
+                        "event loop like this, you can disable this log message by specifying the " +
+                        "-Dcom.linecorp.armeria.checkFutureJoinOnEventLoop=false system property",
+                        new IllegalStateException("Blocking event loop, don't do this."));
```

Comment:
- Perhaps we should log only once for each thread? Could keep a weak set of `Thread`s?
- What about referring to the URL of some documentation which explains this well?

Reply:
Think we might have to write that documentation 😄